openedx-unsupported/registrar
# Implement V1 program enrollment GET

### About
Team: @edx/masters-neem 
Issues: [EDUCATOR-3987](https://openedx.atlassian.net/browse/EDUCATOR-3987) and [EDUCATOR-4111](https://openedx.atlassian.net/browse/EDUCATOR-4111).
Necessary configuration changes: [Configuration PR](https://github.com/edx/configuration/pull/5095) and [Internal PR](https://github.com/edx/edx-internal/pull/841)

This PR does a few things:
* Adds a V1 endpoint for getting program enrollments in a JSON/CSV file
* Adds a "job" system on top of django-user-tasks
* Adds a V1 endpoint for retrieving job statuses
* Uses Boto3 to upload files to Amazon S3 and provide signed URLs for limited-time access
* Uses the Moto library to mock out Boto3 in unit tests
* Uses basic FileSystemStorage for storing generated files when developing locally

### An open issue

For local development, I'm not sure how to return an absolute URL to a job-generated file. I'm thinking that we don't worry about this for now, and revisit it when we address the greater issue of having a configured `BASE_URL`.

### Manually testing
First, create a program in Registrar's Django admin.

Then, mock out the LMS get enrollments API by adding an early return to `_get_all_paginated_results` in registrar/apps/enrollments/data.py:
```python
def _get_all_paginated_results(url, client=None):
    """
    Builds a list of all results from a cursor-paginated endpoint.

    Repeatedly performs request on 'next' URL until 'next' is null.
    """
    return [
        {'student_key': 'a', 'status': 'enrolled', 'account_exists': True},
        {'student_key': 'b', 'status': 'pending', 'account_exists': False},
        {'student_key': 'c', 'status': 'canceled', 'account_exists': True},
    ]
    # ... the rest of the function
```

Next, hit http://localhost:18734/api/v1/programs/{program_key}/enrollments/ with the key of the program you created.

It should return 202 and a job ID and URL. Hit the job URL. It should return a Succeeded status and a result URL.

The result URL is actually a path relative to the host. So, append the path (e.g., `/media/job-results/abcd...json`) to `http://localhost:18734/`, which should give you the JSON file.

Finally, repeat the process with `?fmt=json` and `?fmt=csv` appended to the enrollments URL.
